compiler/circle-mpqsolver/src/bisection/ErrorApproximator.cpp

```diff
@@ -0,0 +1,361 @@
+ * https://github.com/Samsung/ONE/pull/10170#discussion_r1065371638 for details.
+ */
+float approximate_conv(const CircleNode *node)
+{
+  float volume_W_A_err = 0.f;
+  {
+    // activation min-max values
+    float ci_min = 0.f;
+    float ci_max = 0.f;
+    get_min_max_ifm_values(node, ci_min, ci_max);
+
+    // channel-wise min, max
+    std::vector<float> min_values;
+    std::vector<float> max_values;
+    cal_minmax_per_channel(node, min_values, max_values);
+
+    // ranges  = (max_values - min_values)
+    std::vector<float> ranges;
+    std::transform(max_values.begin(), max_values.end(), min_values.begin(),
+                   std::back_inserter(ranges), std::minus<float>());
+
+    // maximal weight value across all channels
+    float w_max = 0;
+    {
+      for (size_t i = 0; i < max_values.size(); ++i)
+      {
+        w_max = std::max(w_max, std::abs(max_values[i]));
+        w_max = std::max(w_max, std::abs(min_values[i]));
+      }
+    }
+
+    // total weight quantization error across all channels
+    // so maximal error of quantization is ~ (max_value - min_value) / 255
+    // omitting 255 term we get that maximal error of quantization is just its range
+    float sum_err = 0.f;
+    for (auto cur_err : ranges)
+    {
+      sum_err += cur_err;
+    }
+
+    uint32_t adds_per_channel = get_additions_per_channel(node);
+    uint32_t num_of_channels = ranges.size();
+
+    // maximal error introduced by weights quantization (for all channels)
+    volume_W_A_err = sum_err * std::max(::fabs(ci_max), ::fabs(ci_min));
+    // plus total error introduced by activation quantization (for all channels)
+    volume_W_A_err += w_max * num_of_channels * ::fabs(ci_max - ci_min);
+    // scale by volume of adds per channel
+    volume_W_A_err *= adds_per_channel;
+    // scale to get more readable output values
+    volume_W_A_err /= 1.e+6f;
+  }
+
+  return volume_W_A_err;
+}
+
+} // namespace
+
+namespace mpqsolver
+{
+namespace bisection
+{
+
+/**
+ * How Approximate works?
+ *
+ * Currently it works just for convolution layers, but may be generalized for other types as well.
+ * See discussion at https://github.com/Samsung/ONE/pull/10170#discussion_r1042246598
+ * Convolution can be expressed as a matrix multiplication.
+ * While quantizing we introduce quantization error into convolution operand (activations) as well
+ * as into convolution weights. A_q * W_q = (A + q_err(A)) * (W + q_err(W)) = A * W + A * q_err(W) +
+ * W * q_err(A) + q_err(A) * q_err(W), assuming q_err(A) * q_err(W) are negligible as quadratic
+ * terms, we get A_q * W_q ~ A * W + A * q_err(W) +  W * q_err(A) , q_err - quantization error,
+ * W - weight matrix, A - activations from previous layer (IFM), so quantization error of matrix
+ * multiplication can be approximated as A * q_err(W) + W * q_err(A). Estimating its upper bound
+ * we get A * q_err(W) + W * q_err(A) <=
+ * number_of_additions * (A_max * (W_max - W_min) / 255 + W_max * (A_max - A_min) / 255)
```

Comment:
Hmm.. Not for this PR, but I think it would be possible to use layer-wise qerrors collected by profiling (by directly comparing the output feature maps ~ output of `visq`), rather than using approximate errors.

Reply:
> Hmm.. Not for this PR, but I think it would be possible to use layer-wise qerrors collected by profiling (by directly comparing the output feature maps ~ output of `visq`), rather than using approximate errors.

Looks very promising!